src/Multivariate/ProductFun.jl

```diff
@@ -198,7 +198,7 @@ canonicaldomain(f::ProductFun) = canonicaldomain(space(f))
 function canonicalevaluate(f::ProductFun{S,V,SS,T},x::Number,::Colon) where {S,V,SS,T}
     cd = canonicaldomain(f)
     Fun(setdomain(factor(space(f),2),factor(cd,2)),
-                    T[setdomain(fc,factor(cd,1))(x) for fc in f.coefficients])
+                    [setdomain(fc,factor(cd,1))(x) for fc in f.coefficients])
```

Comment:
It turns out that canonicalevaluate() has the same problem of wrong type casting.  Before this fix:
```julia-repl
julia> ApproxFunBase.canonicalevaluate(f, tocanonical(f, 1, 2)...)
1.7886132445101346

julia> f = ProductFun{CosSpace{PeriodicSegment{Float64}, Float64}, CosSpace{PeriodicSegment{Float64}, Float64}, TensorSpace{Tuple{CosSpace{PeriodicSegment{Float64}, Float64}, CosSpace{PeriodicSegment{Float64}, Float64}}, DomainSets.VcatDomain{2, Float64, (1, 1), Tuple{PeriodicSegment{Float64}, PeriodicSegment{Float64}}}, Float64}, Int64}(Fun{CosSpace{PeriodicSegment{Float64}, Float64}, Int64, Vector{Int64}}[Fun(CosSpace(0..2π),[1, 3]), Fun(CosSpace(0..2π),[2, 0])], CosSpace(0..2π)⊗CosSpace(0..2π))
ProductFun on CosSpace(【0.0,6.283185307179586❫)⊗CosSpace(【0.0,6.283185307179586❫).

julia> ApproxFunBase.canonicalevaluate(f, tocanonical(f, 1, 2)...)ERROR: InexactError: Int64(2.6209069176044193)
Stacktrace:
...
```